### PR TITLE
translate-c: Implement flexible arrays

### LINKED ISSUE #8759: Imported C Structs with VLAs
I'm currently using 0.8.0-dev.2192+65d279bc0 if that matters.

**Problem:**

I'd like to use [these C structs from ACPICA](https://github.com/acpica/acpica/blob/c1adb9a2a/source/include/actypes.h#L1378-L1406) in my Zig code. The following code is generated from it:

```zig
// (...)/actypes.h:1382:37: warning: struct demoted to opaque type - has variable length array
pub const struct_acpi_pnp_device_id_list = opaque {};
pub const ACPI_PNP_DEVICE_ID_LIST = struct_acpi_pnp_device_id_list;
pub const struct_acpi_device_info = extern struct {
    // ... fields that aren't important ..
    CompatibleIdList: ACPI_PNP_DEVICE_ID_LIST,
};
pub const ACPI_DEVICE_INFO = struct_acpi_device_info;
```

The warning seemed fair by itself when I first looked at this, but this apparently means that I can't use `ACPI_DEVICE_INFO`/`struct_acpi_device_info` at all:

```
./zig-cache/o/3016e146699c4abaeedfebe9c82141cf/cimport.zig:153:5: error: opaque types have unknown size and therefore cannot be directly embedded in structs
    CompatibleIdList: ACPI_PNP_DEVICE_ID_LIST,
    ^
```

Having a variable amount of data at the end of a data structure seems like a common pattern in data structures dealing with hardware. I've had to write code like [this here for USB EHCI](https://github.com/iguessthislldo/georgios/blob/37d86c0296131b7b5dcfeb0013015f809660ae1f/kernel/platform/usb.zig#L75-L77), among other places to access this kind of data from hardware. I'm fine writing code like that and it seems like I'm not the first person writing Zig to do so: #6008.

**Possible Solutions?**

I'm wondering if, for the sake of better compatibility with C, Zig could only demote the variable length array field itself to an opaque type and allow them at the end of the struct. This would [mirror how VLAs act in C](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html). It might also make previously mentioned situation of using this in normal Zig more readable and reduce the amount of pointer arithmetic, but I think that's a minor advantage if it's one at all. Or maybe this should just should be limited to `extern struct`s if it is better to isolate this behavior?

```zig
const MyType = extern struct {
  length: usize,
  data: opaque {},
};
```

Alternatively it would be better from my perspective as a user if Zig fully supported VLAs in structs, but I'm assuming that's not on the table because normal VLAs have been shot down before: #3952.

**Also**

The actual `cimport.zig` lines with the warning are:

```zig
pub const ACPI_PNP_DEVICE_ID = struct_acpi_pnp_device_id; // /data/development/os/georgios/kernel/platform/acpica/acpica/source/include/actypes.h:1382:37: warning: struct demoted to opaque type - has variable length array
pub const struct_acpi_pnp_device_id_list = opaque {};
```

It's kinda weird the comment is following `ACPI_PNP_DEVICE_ID`, shouldn't there be a newline there or it should be following `struct_acpi_pnp_device_id_list`, which is what it's actually talking about?

### PR DESCRIPTION
**Pull request by ehaas**
Fixes #8759

### CODE REVIEW

**Comment by iguessthislldo** [comment]
Thanks! This was blocking me, so I was just about to ask in the issue if I could take a crack at it if no one else was planning to. I will try this out later today.

**Review by Vexu — APPROVED** [review]

File: src/translate_c.zig [hunk 1]
```diff
@@ -819,6 +819,111 @@ fn transTypeDef(c: *Context, scope: *Scope, typedef_decl: *const clang.TypedefNa
...
+    const return_type_decl = try Tag.var_simple.create(c.arena, .{
+        .name = return_type_name,
+        .init = return_type,
+    });
+    try block_scope.statements.append(return_type_decl);
+    const return_type_ident = try Tag.identifier.create(c.arena, return_type_name);
+
+    const field_index = field_decl.getFieldIndex();
+    const bit_offset = layout.getFieldOffset(field_index); // this is a target-specific constant based on the struct layout
+    const byte_offset = bit_offset / 8;
```
Comment (ehaas) [inline src/translate_c.zig:874]: This division assumes 8 bits per byte; is it worth using `toCharUnitsFromBits` to ensure it works on exotic platforms with different-sized bytes?

Comment (Vexu) [reply src/translate_c.zig:874]: I highly doubt this will ever become an issue.

Comment (tothambrus11) [reply src/translate_c.zig:874]: Yep, zig still hasn't decided to support non-octet bytes
https://github.com/ziglang/zig/issues/7693

**Pull request merged**